changelog/mixin_template_mangling.dd

```diff
@@ -0,0 +1,72 @@
+extern(C) declarations in template mixins now mangle as C symbols when mixed in at global scope
+
+This was already true for string mixins, but since template mixins introduce a new namespace, symbols inside them got mangled as D symbols.
+However, users often use extern(C) inside a mixin template to automatically generate boilerplate code that should be accessible from C.
+
+-------
+// library code
+mixin template WasmEntryPoint() {
+    extern(C) export void _start() {
+        // boilerplate code
+    }
+}
+
+mixin template UseGpuInsteadOfIntegratedGraphics() {
+    extern(C) export uint NvOptimusEnablement = 0x00000001;
+    extern(C) export int AmdPowerXpressRequestHighPerformance = 1;
+}
+
+// application code
+mixin WasmEntryPoint;
+mixin UseGpuInsteadOfIntegratedGraphics;
+
+static assert(_start.mangleof == "_start");
+static assert(NvOptimusEnablement.mangleof == "NvOptimusEnablement");
+-------
+
+Previously, _start would be mangled as _D9onlineapp8__mixin46_startUkZv and users had to manually add pragma(mangle, "_start") or use a string mixin instead.
+With the new behavior this is not necessary anymore.
+extern(C++) declarations and mixed in member functions or nested functions are unaffected.
```

Comment:
Why are `extern(C++)` declarations unaffected ?

Reply:
Ah I see the same question was raised. From a user PoV, this is a very annoying inconsistency, and I expect that someone will raise a bug report about it at some point. Could you explain why namespaces are problematic ?

Reply:
I'm not sure what 'inconsistency' or 'problems with namespaces' you are talking about. Let me clarify:

Before this PR:
`extern(C++)` declarations are always mangled externally, other `extern()` declarations are mangled externally when they are _directly in_ global scope.

After this PR:
`extern(C++)` declarations are always mangled externally, other `extern()` declarations are mangled externally when they are _accessible from_ global scope.

- Why the change? Because users expect global mixin templates to count as global scope just like string mixins, but technically it isn't (it's a new scope that's automatically imported in global scope).
- Why is `extern(C++)` always mangled externally? Because C++ mangling supports overloads and namespaces just like D mangling, and it is possible I want to link my D struct member function to a C++ struct member function, so I need C++ mangling there.
- Why is `extern(C)` (/Windows/Pascal/ObjC) not mangled externally when in a function? Because C has no member functions so it is impossible to want your D struct member function to link to a C struct member function, so it is assumed you want the C ABI only but still use D mangling.
- Why not disallow extern(C) member functions then? See https://github.com/dlang/dmd/pull/7229
- How about nested `extern(C)` functions like [issue 15843](https://issues.dlang.org/show_bug.cgi?id=15843)? It is possible to change that, but it will likely break existing code defining callbacks that way.

If those weren't your questions, please tell me.

Reply:
I changed the changelog a bit to reflect that C++ already behaved as expected, can you approve it now or request changes?